M2/Macaulay2/m2/methods.m2

```diff
@@ -1,159 +1,262 @@
-	  );
-     methodFunction)
-
+-- TODO: handle this in the interpreter, the same way that it is handled for function closures
 chkopt0 := k -> if not ( instance(k, Symbol) ) then error "expected SYMBOL => VALUE"
 chkopt  := o -> if not ( class o === Option and #o === 2 and instance(o#0, Symbol) ) then error "expected SYMBOL => VALUE"
 chkopts := x -> if class x === OptionTable then scan(keys x,chkopt0) else if class x === List then scan(x,chkopt) else error "expected list of optional arguments"
+badopts := x -> "options must be declared as a list, an option table, or as true for arbitrary options"
 
-SingleArgWithOptions := (opts,outputs) -> (
-     -- chkopts opts;
-     local methodFunction;
-     class' := if outputs then identity else class;
-     if opts === true then (
-	  methodFunction = opts >> 
-	  o ->
-	      arg -> (
-	       -- Common code for every method with options, single argument, with Options => true
-	       f := lookup(methodFunction, class' arg);
-	       if f === null then noMethodSingle(methodFunction,arg,outputs) else 
-	       if #o === 0 then f arg else f(o,arg)
-	       );
-	  methodFunction
-	  )
-     else (
-	  if instance(opts, List) then opts = new OptionTable from opts;
-	  methodFunction = opts >> 
-	  o ->
-	      arg -> (
-	       -- Common code for every method with options, single argument, not Options => true
-	       f := lookup(methodFunction, class' arg);
-	       if f === null then noMethodSingle(methodFunction,arg,outputs) else (f o) arg
-	       );
-	  methodFunction
-	  )
-     )
+-----------------------------------------------------------------------------
+-- MethodFunction* type declarations and basic constructors
+-----------------------------------------------------------------------------
 
-BinaryWithOptions := (opts,outputs) -> (
-     -- chkopts opts;
-     error "associative methods with options not implemented yet"
-     )
+MethodFunction = new Type of CompiledFunctionClosure
+MethodFunction.synonym = "method function"
+
+-- TODO: are these two really useful?
+MethodFunctionSingle = new Type of FunctionClosure
+MethodFunctionSingle.synonym = "method function with a single argument"
+
+MethodFunctionBinary = new Type of FunctionClosure
+MethodFunctionBinary.synonym = "associative binary method function"
+
+MethodFunctionWithOptions = new Type of FunctionClosure
+MethodFunctionWithOptions.synonym = "method function with options"
+
+-----------------------------------------------------------------------------
+-- helpers for method
+-----------------------------------------------------------------------------
+-- from interpreter:
+-- newmethod1
+-- newmethod1234c
+
+notImplemented = x -> error concatenate between(" ", splice {x, "not implemented yet"})
+
+-----------------------------------------------------------------------------
+
+-- see fold.m2
+-- TODO: add Left associative vs Right associative methods
+foldL := (f, x, L) -> (scan(L, y -> x = f(x, y)); x)
```

Comment:
`fold` is a method that is defined in a later source file. I can't use a method to construct methods!! It would be good to put `foldL` and `foldR` in the interpreter, perhaps.

Reply:
Let me be the one to "resolve" my conversations.

The method function `fold` can be used to make a method if the method isn't called before `fold` is defined.   It can probably work.  That would be good because it would eliminate duplication of code.

Reply:
I understand that, but a big point of this PR was to optimize methods as much as I could without changing the interpreter, and using a method call in the function that evaluates another method call slows things down quite a bit.

I'm surprised that you're concerned about duplication of literally one line of code when there are huge repetitions all over the place, but do you prefer if I used `foldL` in definition of `fold`?

Feel free to turn this conversation into a new issue for someone to implement a basic `foldL` function in the interpreter, which is then used in both `methods.m2` and `fold.m2`. This is my preferred solution, but I don't want to deal with the interpreter in this PR.

Reply:
I resolved duplicated code.